### PR TITLE
[DistillationTrainer refactor] Support tool calling

### PR DESCRIPTION
**Pull request by qgallouedec**
## What

> [!IMPORTANT]
> For reviewers: it's mostly copy-pasted from GRPO, nothing really new here!

Adds tool-calling support to `DistillationTrainer`, mirroring GRPO. The student generates multi-turn tool-calling trajectories on-policy; tool-result tokens are masked out of the JSD loss.

- `max_tool_calling_iterations` config field (caps tool turns; `None` = unlimited)
- `tools` constructor param + `_tool_dict`, `supports_tool_calling` / response-schema / training-chat-template setup
- `_tool_call_loop`, `parse_response`, `_get_tool_suffix_ids` in the generation path
- `tool_mask` loss masking: `loss_mask = completion_mask * tool_mask`
- `test_train_with_tools` (mocked `generate`, asserts `tools/call_frequency` and `tools/failure_frequency`)

All mirrored blocks are reproduced verbatim from GRPO.

## Deferred (future PR)

Async tools, environments, and per-token logprob capture — none exist on a stable base yet.

## Stacking

Stacked on #6650 (VLM). Re-base to `main` once #6650 lands.

Part of #6449.

---

> [!NOTE]
> **Medium Risk**
> Touches core on-policy generation, loss masking, and VLM forward kwargs in `DistillationTrainer`; behavior is largely ported from GRPO but multimodal tool loops add integration risk.
> 
> **Overview**
> **Adds on-policy tool-calling to `DistillationTrainer`**, aligned with GRPO: pass a `tools` list, run a multi-turn generate→execute→regenerate loop, and distill only on student-generated tokens by masking tool-result tokens out of the JSD loss (`loss_mask = completion_mask * tool_mask`).
> 
> `DistillationConfig` gains **`max_tool_calling_iterations`** to cap tool turns. Init wires response parsing, prefix-preserving chat templates, transformers ≥5.0 checks, and **rejects async tools**. Logging adds **`tools/call_frequency`** / **`tools/failure_frequency`**, and token/completion length metrics note they exclude tool tokens when tools are enabled.
> 
> **VLM paths** merge tool-returned images into generation and the forward pass (including `mm_token_type_ids` / truncation safety). Docs add an **Agent Training** section for distillation; GRPO docs document the same iteration cap.
> 
> Tests cover mocked multi-turn tool calls (success/failure metrics) and VLM tool responses (image and text).
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 8ed5196d59670e2f2c630ca7c1c19036b1e55902. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/trainer/distillation_trainer.py [hunk 1]
```diff
@@ -1135,8 +1566,69 @@ def _generate_and_score_completions(self, inputs: list[dict[str, torch.Tensor |
...
+        loss_mask = completion_mask if tool_mask is None else completion_mask * tool_mask
+        num_items_in_batch = self.accelerator.gather(loss_mask.sum()).sum()
 
-        num_items_in_batch = self.accelerator.gather(completion_mask.sum()).sum()
+        num_images = [len(img_list) if img_list else 0 for img_list in images] if images is not None else None
+
+        # Get forward_kwargs for models with multimodal inputs.
+        if images is not None:
+            prompts_text = [
+                apply_chat_template({"prompt": prompt}, self.processing_class, **self.chat_template_kwargs)["prompt"]
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/trainer/distillation_trainer.py]: Valid, `_tokenize_prompts` passes tools, this path doesn't. Fixed by passing `tools=self.tools or None`, matching GRPO.

File: trl/trainer/distillation_trainer.py [hunk 2]
```diff
@@ -1007,6 +1119,239 @@ def _generate_single_turn(self, prompt_ids, images, multimodal_fields):
...
+            merged_images = images
+            if any(imgs for imgs in tool_images):
+                if merged_images is None:
+                    merged_images = [imgs if imgs else None for imgs in tool_images]
+                else:
+                    merged_images = [
+                        (existing or []) + new for existing, new in zip(merged_images, tool_images, strict=True)
+                    ]
+            loop_images = [merged_images[i] for i in idxs_with_tool] if merged_images else None
+            if multimodal_fields:
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/trainer/distillation_trainer.py]: Correct `pixel_values`/`image_grid_thw` are packed per image, not per row. Ported GRPO's split/select/unsplit block verbatim.

File: trl/trainer/distillation_trainer.py [hunk 3]
```diff
@@ -1007,6 +1119,239 @@ def _generate_single_turn(self, prompt_ids, images, multimodal_fields):
...
+                loop_multimodal_fields = {}
+                for k, v in multimodal_fields.items():
+                    selected = [v[i] for i in idxs_with_tool]
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/trainer/distillation_trainer.py:1342]: Right, the post-tool generation reuses pre-tool fields. Added the `image_processor` reprocessing branch and the `has_tool_images` token-type rebuild, mirroring GRPO.

File: trl/trainer/distillation_trainer.py [hunk 4]
```diff
@@ -1211,8 +1566,16 @@ def _generate_and_score_completions(self, inputs: list[dict[str, torch.Tensor |
...
+            tool_mask = [torch.tensor(mask) for mask in tool_mask_list]
+            tool_mask = pad(
+                tool_mask, padding_value=1, padding_side="right", pad_to_multiple_of=self.pad_to_multiple_of
+            ).to(device=device)
+        else:
+            tool_mask = None
 
-        num_items_in_batch = self.accelerator.gather(completion_mask.sum()).sum()
+        loss_mask = completion_mask if tool_mask is None else completion_mask * tool_mask
+        num_items_in_batch = self.accelerator.gather(loss_mask.sum()).sum()
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/trainer/distillation_trainer.py:1651]: Agreed `tool_images` isn't propagated to the loss path at all. Threaded it through and added GRPO's tool-image `forward_kwargs` branch.

**Comment by qgallouedec** [comment]
Deep review done before merging: cursor's vllm_mode catch was real (AttributeError with use_vllm + tools), fixed in 8ed5196d along with gating the decode on tools and doc alignment; the other bot findings were fixed earlier or exist verbatim in GRPO.
Remaining findings are GRPO-shared (multi-GPU tool-loop collectives, VLM chat_template re-render, tools or None guard), filing as separate issues. Review request open 12 days, self-merging.
